### ISSUE TITLE
What to do with partially sold pairs?

### ISSUE THREAD

**Issue by yazeed** [freqtrade/freqtrade#3009]
This is an issue to discuss what can be the best way to go about handling partially sold pairs, as I've noticed in the code there's a TODO part. Once established a consensus on an approach, a PR can be made as a follow-up.

`# TODO: figure out how to handle partially complete sell orders`

My opinion on this is that possibly there could be some mechanism that either:
- Triggers based price slipping away from partially sold items
- Triggers based on sell order timing out

Accordingly, it can proceed to do the following:
- _Cancel existing sell order on exchange, so that freqtrade can decide what to do with remaining amount according to stoploss/minimal_roi/trailing conditions in strategy._
- Modify trade in persistence to fulfilled amount, mark it as closed, and calculate profit according to fulfilled amount and timestamps, etc...
- Create a duplicate of the trade in persistence with the remaining amount.
- ~~Cancel existing sell order on exchange, so that freqtrade can decide what to do with remaining amount according to stoploss/minimal_roi/trailing conditions in strategy.~~

Would love to hear your thoughts.

**Comment by xmatthias** [comment]
>Cancel existing sell order on exchange, so that freqtrade can decide what to do with remaining amount according to stoploss/minimal_roi/trailing conditions in strategy.

Well this seems pointless - the order did time out - so cancelling the order is "automatic" to me ... handling would happen after it's cancelled.

---

Assuming the order did time out and was only partially sold....
you'll need to consider a few cases - which makes handling of this very difficult.

Surprisingly, while we got this TODO comment in the code since years, I've not seen any issues related to half-filled sell orders... (as in - never)...

Either way, when taking this topic, we need to consider the following cases:

* little sold (almost full trade is still present)
* halfway sold (remaining balance > minimum tradable amount)
* mostly sold (remaining balance < minimum tradable amount)

I suspect the first 2 cases can be handled in a similar way... the 3rd one however is the difficult one.

If you create a new trade for the 3rd case, you'll run into "unable to sell" errors immediately.

Unfortunately, the amounts tied up in this can be a few USDT easily (binance uses 10USDT as minimum order size [source](https://www.binance.com/en/trade-rule)) - so a unfortunate order can remain stuck with 9.5USDT worth of Crypto currency - which it's not able to convert to USDT again (unless you buy another 10 USDT - then you can sell the 19.5 USDT worth of USDT easily).
Because of this - simply ignoring these could lead to "vanishing stakes" - where your stake-currency becomes lower and lower, while the bot is making profits. 

The easy answer for binance users is "convert to BNB" - which is what Binance support would tell you - but what if you're not on binance, but have the same problem on Kraken or Bittrex (or any other exchange)? handling should work identically for all - but none of the other exchanges has a "convert dust to ..." button ...

**Comment by yazeed** [comment]
> mostly sold (remaining balance < minimum tradable amount)

For this specific case, I think we can have freqtrade keep the sell order as per the existing behaviour.

We'd still benefit from having the first 2 cases you've mentioned covered. This depends on how we decide to trigger the suggested behaviour (time based or price slipping away threshold based)

One risk of not having the above proposed feature is essentially specific to sell orders (when a sell order is above pair's current price), potentially having a majority of the trade (significant amount) depreciate in value overtime, without allowing freqtrade to conduct any strategy logic on it, preventing from freqtrade to go back to use a strategy's trailing stop or stoploss logic.

**Comment by xmatthias** [comment]
> This depends on how we decide to trigger the suggested behaviour (time based or price slipping away threshold based)

to be honest, i'm not sure what you mean with this, and i think you're mixing two things ...
One thing is the timeout trigger - which we already got (identically for buy and sell orders) based on timeout, and soon based on user callbacks (#2872 ). If you wanna use it by difference to current price - that can be easily done in the callback (as the sample there shows).

The other thing is what i understand this issue is for - which is handling partially filled sell-orders (which is something we're not handling at the moment by simply leaving these open).

To me, the moment you get to anything that's relevant for this issue - the decision that this sell-order should be cancelled has already been taken.

Keep in mind, closing of unfilled sell orders does work fine as long as the order did not start to fill.

**Comment by hroff-1902** [comment]
My old idea was to add a new option, like `--ignore-leading-incomplete-trades` (with some shorter name), which will exclude partial trades from backtesting results. For both backtesting and hyperopt.
Rationale: to avoid the skew in results caused by incomplete trades, that can significantly modify the total result, since they are forcibly sold at the end of the period.
A variant of this: another option (or 'mode') to ignore only incomplete losses (i.e. emulate force sell for profitable trades, fixing profit at the end of the period). Both modes (plus third -- current behavior, include incomplete trades into consideration) can be implemented and available for backtester.

Should be easy for implementation -- just remove the trades with appropriate sell reason after backtesting session.

**Comment by devilsinthedetail** [comment]
I'm not sure if this is an issue for a lot of people but I notice that quite often I get partially filled orders which: 
(1) Spam the telegram bot every 200ms with: "Cancelling Open Sell Order for XXX/TTT. Reason: partially filled - keeping order open"
(2) can not be forceselled from telegram
Is there another way of resolving this without manually cancelling the trade on the exchange?

**Comment by xmatthias** [comment]
We'll work on a fix for the spamming soon.

For the forceselling - this part is difficult to solve as it has many quirks / edge cases in which you'll end up stuck with what you have if you cancel the order.

for example:
assume asset XXX has minimum trade size of 10 (so every order must involve at least 10 XXX).
We own 20 XXX and open a sell-order.

There are the following cases that i know of (just out of memory, without thinking hard):
* the order is fully filled
* the order is not filled at all
* 5 XXX are sold before the timeout
* 15 XXX are sold before the timeout

Now the first 2 cases are easy to handle (and are handled correctly).

The 3rd case should be handlable by opening a new sell-order for the remaining 15 XXX - but assuming the detection is somehow even so slightly wrong - then it'll fall into the below category.

However the last case is the problematic one.
If this order is cancelled with the 5 XXX remaining, these 5XXX cannot be sold anymore 
If you were to manually solve this, you'd have to buy 10 XXX, and then sell the 15XXX you now own again together.
Due to fees this will incur (selling now took you 4 trades), this would most likely wipe out most of the profit you made.

What makes this even harder is that it's not reliably testable - with test trades (not done by a regular strategy), i usually try to stay around the lower limits of the minimum tradable.
Most of these trades make losses - since i buy them when i start debugging, and sell them whenever i need to, so i will not want to invest a huge sum.

Since these trades then involve low quantities - it's very difficult and rare to encounter partial sells (however to be honest, i almost never have partial fills on my real bots (which use higher amounts) either - but i (my strategy) tries to stay away from low-volume coins.

**Comment by devilsinthedetail** [comment]
Thanks for the explanation. 
So wouldn't it cover most of the use cases to just close the order and go back to "normal" behavior, so that next loop (given a sell signal) it can open a new sell with the remaining amount. This will only run into problems when going below the minimum trade size but this is a problem which can not be circumvented automatically in any case (not that easily at least, buying more, then selling could be one solution)

**Comment by devilsinthedetail** [comment]
I suggest to use a variable `keep_partial_sell_open` which could be introduced probably in [this function](https://github.com/freqtrade/freqtrade/blob/5b923877320e0fcd24471663aed1b9e88b39c7bd/freqtrade/freqtradebot.py#L943) which if True does everything like now, but if False cancels the trade. This would at least solve one of the problems?

**Comment by xmatthias** [comment]
i don't think we'll need a variable for it if the checking of "is > min tradable amount" is done correctly.

The change should be pretty limited (mostly within `handle_timedout_limit_sell()`)...

Feel free to go for it if you have the time and will - we can support if needed (with tests most likely - which is a pain for many new contributors).

**Comment by devilsinthedetail** [comment]
I had a look at the code, still feel a bit unsure about making a pull request, maybe I can ask some questions here. I adjusted the code in the following way, handling of partially sold pairs at the beginning so that all end statements like trade.close_rate etc. have to be only handled once. 

```python 
def handle_timedout_limit_sell(self, trade: Trade, order: Dict) -> str:
    """
    Sell timeout - cancel order and update trade
    :return: Reason for cancel
    """
    # trade is partially filled but can not be traded
    if order['remaining'] <= min_tradable_amount and order.get('filled') > 0.0:
        return 'partially filled - keeping order open (remaining amount < trading limit)' # Should maybe be a warning to handle this manually

    # trade is partially filled but can still be traded. Cancel here and wait for next sell signal. 
    elif order['remaining'] > min_tradable_amount and order.get('filled') > 0.0:
        reason = "partially filled - cancelling order (remaining amount > trading limit)"
        self.exchange.cancel_order(trade.open_order_id, trade.pair)
        logger.info('Sell order %s for %s.', reason, trade)

    # if trade is not partially completed, just cancel the trade
    elif order['remaining'] == order['amount'] or order.get('filled') == 0.0:
        if not self.exchange.check_order_canceled_empty(order):
            reason = "cancelled due to timeout"
            # if trade is not partially completed, just delete the trade
            self.exchange.cancel_order(trade.open_order_id, trade.pair)
            logger.info('Sell order %s for %s.', reason, trade)
        else:
            reason = "cancelled on exchange"
            logger.info('Sell order %s for %s.', reason, trade)

    trade.close_rate = None
    trade.close_rate_requested = None
    trade.close_profit = None
    trade.close_profit_abs = None
    trade.close_date = None
    trade.is_open = True
    trade.open_order_id = None

    return reason
```
Some questions about that: 
 

1. Where do I get min_tradable_amount from? Is there some field in trade for that?
2. From my experience roughly 5% of trades end up in the partially filled case and none of the ones I had ended up in the partially filled and non tradable case (but its only running since a week so its just a matter of time with low volume pairs). If it is cancelled in the first elif then the sell signal will just come back in and the remaining amount is sold? 
3. Regarding the remaining amount, do we have to set this manually so that the rest of the code is not confused, i.e. sth like order['amount'] = order['remaining']?

**Comment by xmatthias** [comment]
I think you'll need to indent the last section (upating `trade`) into the last if.

1) that's the big problem - we do have `_get_min_pair_stake_amount()` - the problem is, it doesn't just depend on the amount, but also on the price (there's 2 limits that need respecting on most exchanges, amount and cost).
so even though you might be fine cancelling the order at this point - 2 minutes later it could be below the cost level (as price dropped).

2) at the moment (with your above code) - no. But this is the next problem - if it filled halfway - you can't reduce the amount to that (that works on buying, but not on selling). If you reduce the amount - then you'd be hiding the profit/loss you've already made, as you tell the bot to "you never had more than 10XYZ" - even though the initial trade was for 100 XYZ - so this will influence the whole profit / loss calculation (you could have made 5% profit on the first 90 XYZ - and the last 10 produce -2%. By updating trade_amount to the 10 XYZ - all reports would only show the -2% - although the trade overall was positive.


3) yes (kindof) - although i'm carefull with using 'remaining' - as that may or may not be set correctly (FTX does set it to 0 on cancelled orders).



To summarize: As i've said above - it's not "that easy" ... to handle this :)

**Comment by Retcorp** [comment]
Hi Folks,

Long time reader, first time poster....Loving your work so far btw :-)

To weigh in on this discussion, I've recently sent a strategy live on Kraken that looked great in backtesting over several months (but as we know, backtesting obviously assumes there are always buyers readily available), but when live, it has got hung up on selling some low liquidity pairs that have very partially filled on the sells, then produced the spamming cancellation message as discussed above, meanwhile, the value has slid downwards and the bot has not had the mechanism to cancel the order and chase the price and offload the remaining. In the end, I've gone direct to the exchange and cancelled the orders myself and then "fire sold" them before they became worthless. I've just delved into the code which lead me to this thread, so have a vested interest in it as it's happened several times to me in the short time I've been live. 

I'm assuming that my "exchange meddling" has annoyed the app, as it then appears to be out of sync and doesn't seem to understand where it is any more. Have I set the app up incorrectly for scenarios such as this, or is there no way that it can "catch up" by getting hold of my order history? I entirely apologise if this is somewhat newbie in nature. :-)

**Comment by xmatthias** [comment]
@Retcorp you'll need to use the [sql cheatsheet](https://www.freqtrade.io/en/latest/sql_cheatsheet/) and update the trades you manually sold on the exchange.

**Comment by AngryByte** [comment]
Hello,

I think issue 1 is not limited to partially filled order, with a small enough stack we can fall into the same trap with a new order if the price * amount falls beyond the minimum allowed on the exchange.

for item 2, I agree that that the profit calculations will suffer if we sell the same order with 2 different rates.
I suggest the following as an alternative:
1. calculate unfilled = order['amount'] - order.get('filled')
2. modify the original order in the database so that "amount" = filled.
3. cancel the order on the exchange and close the trade in the database.
4 create a new "virtual" trade with the original open_rate and order['amount'] = unfilled and a buy fee of 0
5 the virtual order will now wait for a sell signal at a new rate, roi, or stop_loss

if this is a direction you're willing to take, I will try to write some code to get started.

ultimately, it would be best to restructure the database so that multiple sells can be linked to a single buy.

I just got started with freqtrade and I absolutely love it, great job!


> I think you'll need to indent the last section (upating `trade`) into the last if.
> 
> 1. that's the big problem - we do have `_get_min_pair_stake_amount()` - the problem is, it doesn't just depend on the amount, but also on the price (there's 2 limits that need respecting on most exchanges, amount and cost).
>    so even though you might be fine cancelling the order at this point - 2 minutes later it could be below the cost level (as price dropped).
> 2. at the moment (with your above code) - no. But this is the next problem - if it filled halfway - you can't reduce the amount to that (that works on buying, but not on selling). If you reduce the amount - then you'd be hiding the profit/loss you've already made, as you tell the bot to "you never had more than 10XYZ" - even though the initial trade was for 100 XYZ - so this will influence the whole profit / loss calculation (you could have made 5% profit on the first 90 XYZ - and the last 10 produce -2%. By updating trade_amount to the 10 XYZ - all reports would only show the -2% - although the trade overall was positive.
> 3. yes (kindof) - although i'm carefull with using 'remaining' - as that may or may not be set correctly (FTX does set it to 0 on cancelled orders).
> 
> To summarize: As i've said above - it's not "that easy" ... to handle this :)

**Comment by xmatthias** [comment]
The database should allow multiple sells in theory already (we do have a orders table now) - at least with some modifications on how the database is used.
The orders table is currently not used for tracking buys / sells - but only to refind "lost" orders (which was the most urgent problem).

> 4 create a new "virtual" trade with the original open_rate and order['amount'] = unfilled and a buy fee of 0

I don't think this is correct - fees are in ratios (and cost - but the cost column is not really used) - if you reduce the original amount (say from 200 "whatever" to 100) - then you still payed 0.001% Fee - on both (virtual) trades of 100 - so setting this to 0 will make it seem like you only paid half of the fee - which is not correct.

---

One of the more important things i think is correctly identifying trades where the remaining amount is below tradable amount.
Say the bot bought something for 100 USDT - and sold 95% of the trade - it would not be able to create a new order for the remaining 5 USDT (minimum order amount is usually around 10 USDT on binance) so the "virtual" trade you created will remain in your bot forever - as it's impossible to sell. In that case, i'd prefer to keep the order open, as it'll eventually come back to be sold (as long as the order isn't cancelled and replaced, it's not a problem).

**Comment by werkkrew** [comment]
I don't know if this is exactly on topic (but I think it is).

I have noticed after about a week on Kraken trading on some pretty low volume alts I have an untradably low balance of about a dozen coins at the moment.

Ideally, the way I would want the bot to deal with this is to have some mechanism where when the bot trades back into that coin, it is allowed to sell 100% of what I have instead of trying to sell exactly what it bought.

It might be nice to have a configuration area where you can say something along the lines of a whitelist/blacklist whereby you tell freqtrade which coins it can sell all of vs which to obey exact matches between buys and sells on.

For example:

1. Stake currency is ETH and XLM/ETH is on whitelist 
2. A buy -> sell round trip happens and for whatever reason I am left with 0.00001 XLM which is below the trade threshold
3. At some point in the future the bot buys 0.005 XLM, now my balance is 0.005001 XLM
4. When the bot tries to sell, instead of selling the 0.005 it bought, it sells the full XLM balance

In some cases you would not want to do this. If for example my kraken holdings include some things I dont want Freqtrade to dip into, for example lets say I am holding BTC, ZEC and DOGE and I don't want freqtrade to sell 100% of that next time it enters an ETH/ZEC trade (following the above example where ETH is the stake currency is ETH/ZEC is a whitelisted pair.

Perhaps there can be a config item like:

```
    "wallet": {
        "sell-all": True,  
        },
        "coin_blacklist": [
            "ZEC",
            "BTC",
            "DOGE"
        ]
    },
```
Obviously this example could use some work but the idea is "allow freqtrade to sell 100% of the balance of any coin it wants to after it enters a trade, except these"

**Comment by xmatthias** [comment]
to me, that's honestly a terrible idea, and will break everyone's workflow if they do manual trading on the same wallet.
It'll work if you HODL these coins for sure, but it'll completely break all other usecases (even running multiple bots on the same exchange, as then each bot would sell the other bot's coins).

I'd be limited with my manual trading to the preconfigured list - however, when manually trading, you usually look for the right patterns / whatever - and then buy (usually not caring about the coin per se) - so all coins are will be trade depending on the trade situation..

it would also break all profit calculations - assumming it buys 200 XLM for a (say 0.5$) - and then sells 220 XLM for the exact same price (0.5$) - it didn't quite make 10% profit, but it actually made a loss (fees!) - and it has no idea if the 20 XLM were bought at 0.1$ (great profit) - or 1$ (50% loss).

---

I don't quite understand why you're facing this either - i do trade on kraken as well - and i've never had such a situation (i don't own ANY non-tradable balance, and never did anything in this regards).

Freqtrade will not cancel a partial sell order - and i'm honestly not sure if kraken allows you to place a sell order for 0.005001 (manually or through bot) - or if it'll just round it to 0.005.

either way - it's not related to this topic either - as partially sold pairs are something different to what you're seeing.

**Comment by epigramx** [comment]
I believe this area is more of a problem to be allowed to exist as it is compared to having a basic solution first before developing a more complex list of options for all possible problems people have with the basic options.

Many people will not mind hodling a small amount because ..it is a small amount and some exchanges may allow the small amount to be exchanged with a fee.

As a result an optional feature could a) close the partial and count it b) retry with a new order # c) if it can't sell because it's too low just hodl (and delete it).

**Comment by barisengez** [comment]
I have observed a bad scenario about partially sold trades. I ll ignore untradable amount subject and speak only for binance for the sake of this scenario.

it partially sells the trade lets say 10% then the price keeps dipping and the 90% of the trade goes into a big dip with no updated sell price thus never filling the limit order. sell_timeout doesnt take into effect so the partial sell is kept indefinitely here. the situation is not resolved unless manuel intervention of cancelling sell limit order from exchange.

1- can we workaround this by using order_time_in_force = fok on sell side so that sell price is updated in case not filled fully until sell timeout?
2- there is a warning here that order_time_in_force is not yet ready for sell side. is it still the case?
"!!! Warning This is an ongoing work. For now it is supported only for binance and only for buy orders. "
https://github.com/freqtrade/freqtrade/blob/develop/docs/configuration.md#understand-order_time_in_force

**Comment by xmatthias** [comment]
No, it's not uptodate, and works for sell-orders as well.
i've updated the docs to reflect that.

**Comment by SmartManoj** [comment]
Is this issue about this?
![image](https://user-images.githubusercontent.com/7231077/136694213-72bf3f1a-7f9c-483a-a707-96ddc2005fc0.png)

**Comment by SmartManoj** [comment]
https://github.com/freqtrade/freqtrade/blob/5b923877320e0fcd24471663aed1b9e88b39c7bd/freqtrade/freqtradebot.py#L986

Exchange giving lesser amounts?

**Comment by xmatthias** [comment]
No it is not. It's about half-filled orders (where you place an order for to sell 100 coins, and the exchange only fills 50 coins).

Your problem looks to be fee-related - which can easily be mitigated by having BNB in your account.

**Comment by gaugau3000** [comment]
Hi,

Proposal:

Given the limit sell order is partially filled and the amount is upper of minimum trade amount when the sell order timeout is reach then i cancel the order and place a new one.
Given  the limit sell order is partially filled and the amount is lower of minimum trade when the sell order timeout is reach then we keep this order open and close the trade is database (residual consider as a loss) and sent a warning to the user "order still open but trade close to allow freqtrade to be able to trade..."

Freqtrade still continue to trade, it take the pessimistic scenario of loss of residuals.

I don't see elegant way to solve this problem.

I Have some trades lock in sells during days if the price not come back to previous price or do manual action.

Have a nice day.

**Comment by xmatthias** [comment]
> Given the limit sell order is partially filled and the amount is upper of minimum trade amount when the sell order timeout is reach then i cancel the order and place a new one.

This part seems sensible

> Given the limit sell order is partially filled and the amount is lower of minimum trade when the sell order timeout is reach then we keep this order open and close the trade is database (residual consider as a loss) and sent a warning to the user "order still open but trade close to allow freqtrade to be able to trade..."

I disagree with this.
I'd want freqtrade to keep in the trade here - assuming price goes back to the sell-order, i'd want freqtrade to properly track "how it ended".
Assuming the 10$ (or however much) are lost will pull your win/loss ratio shown in freqtrade way down for no good reason (other than bad pricing decisions) ... 
(you've won that trade, and once the remainder fills it's a winning trade.

**Comment by gaugau3000** [comment]
> I'd want freqtrade to keep in the trade here - assuming price goes back to the sell-order, i'd want freqtrade to properly track "how it ended".
> Assuming the 10$ (or however much) are lost will pull your win/loss ratio shown in freqtrade way down for no good reason (other than bad pricing decisions) ...
> (you've won that trade, and once the remainder fills it's a winning trade.

Seems a good option. But this trade have to been considered has close for been able to take new trades ? Because if not it will eat you a "slot" of max open trades and may this situation continue during days or month (i have encounter this situation quite often with strategies that make a lot of trades ).

Have a nice day.

**Comment by gaugau3000** [comment]
Discord exchange to follow  possible workflow:

**gaugau :** 
i may try to work on this issue : https://github.com/freqtrade/freqtrade/issues/3009, if you have an opinion with what to do with "residuals" (trade below the tradable amount). Tks. Have a nice WE.
**xmatthias:** 
not really - i think it'll depend between users what they actually want/need - some will want it to just "forget" about this (maybe consider it as "loss"?) - others will want to keep the order open, so it at least has a chance to fill later ... there's a reason it's a "discussion" labeled issue - as the answer to this is really unclear
**gaugau**
maybe just do the first part at first-> continue place sell orders if total_filled_amout < order_amount - minmum_tradable_amount  and in a second part deal with residuals . The first part is important to preserve capital, the second part is a borderline case with low capital impact and low frequency occurrence. Have a nice day.

**Comment by sckebby** [comment]
Hi, 

Kucoin via API can sell all the Amount / Size of investment by using the MaxQuoteSize,
so for example you have 0.2435 of a coin, if the MaxQuoteSize is 99999, you can place an order with 99999 and all the 0.2435 will be sell without dust remaining,
Does Binance allow to sell a amount Big/Bigger then your investment, to sell ALL your investment as Kucoin API do?

Regards

**Comment by VanessaE** [comment]
I say just time-out after some period, kill the order, and leave it up to the strategy to deal with the remaining amount vs. the exchange's minimum order size, even if that means the strat has to buy some extra in order to be able to sell the remainder.

**Comment by alisalama** [comment]
I too am getting some partially filled sells that are effectively stuck without manual intervention after price drifts away. Using a 1 minute timeout.

I understand the problem of below the minimum order size, but I am seeing 10% partial sells get stuck on BTC (obviously highly liquid) and the other 90% (easily over the minimum threshold) just drain away as the price drifts.

This seems a problem for scaling of the bot. It effectively places an unknown-level artificial profitability ceiling related to downside volatility on the bot/strategy. As your stack increases, the likelihood / frequency of this happening will go up, and as it only happens on the downside the profitability will drop / max drawdown will increase.

For my use case cancellation of them all on timeout would be more than acceptable, but I accept doesn't fits everyones use case.

As a catch all solution, would it be possible to have another strategy callback for a customisation of a cancel or no cancellation of a partial sell? This could default to the current "False" behaviour but could be amended to the "True" if you wish to cancel.  Either the strategy could resolve what to do next, or you could pass the price at which you want a new order for the remaining portion to be placed at too.
